src/ripple/app/consensus/RCLConsensus.cpp

```diff
@@ -0,0 +1,882 @@
+#include <ripple/app/misc/LoadFeeTrack.h>
+
+namespace ripple {
+
+
+RCLConsensus::RCLConsensus(
+        Application& app,
+        std::unique_ptr<FeeVote> && feeVote,
+        LedgerMaster& ledgerMaster,
+        LocalTxs& localTxs,
+        InboundTransactions& inboundTransactions,
+        typename Base::clock_type const & clock,
+        beast::Journal journal)
+        : Base(clock, journal)
+        , app_ (app)
+        , feeVote_ (std::move(feeVote))
+        , ledgerMaster_ (ledgerMaster)
+        , localTxs_(localTxs)
+        , inboundTransactions_{ inboundTransactions }
+        , j_ (journal)
+        , nodeID_{ calcNodeID(app.nodeIdentity().first) }
+        , valPublic_ (app_.config().VALIDATION_PUB)
+        , valSecret_ (app_.config().VALIDATION_PRIV)
+{
+
+}
+
+void
+RCLConsensus::onStartRound(RCLCxLedger const & ledger)
+{
+    inboundTransactions_.newRound(ledger.seq());
+}
+
+// First bool is whether or not we can propose
+// Second bool is whether or not we can validate
+std::pair <bool, bool>
+RCLConsensus::getMode ()
+{
+    bool propose = false;
+    bool validate = false;
+
+    if (! app_.getOPs().isNeedNetworkLedger() && (valPublic_.size() != 0))
+    {
+        // We have a key, and we have some idea what the ledger is
+        validate = true;
+
+        // propose only if we're in sync with the network
+        propose = app_.getOPs().getOperatingMode() == NetworkOPs::omFULL;
+    }
+    return { propose, validate };
+}
+
+boost::optional<RCLCxLedger>
+RCLConsensus::acquireLedger(LedgerHash const & ledger)
+{
+
+    // we need to switch the ledger we're working from
+    auto buildLCL = ledgerMaster_.getLedgerByHash(ledger);
+    if (! buildLCL)
+    {
+        if (acquiringLedger_ != ledger)
+        {
+            // need to start acquiring the correct consensus LCL
+            JLOG (j_.warn()) <<
+                "Need consensus ledger " << ledger;
+
+            // Tell the ledger acquire system that we need the consensus ledger
+            acquiringLedger_ = ledger;
+
+            auto app = &app_;
+            auto hash = acquiringLedger_;
+            app_.getJobQueue().addJob (
+                jtADVANCE, "getConsensusLedger",
+                [app, hash] (Job&) {
+                    app->getInboundLedgers().acquire(
+                        hash, 0, InboundLedger::fcCONSENSUS);
+                });
+        }
+        return boost::none;
+    }
+
+    assert (!buildLCL->open() && buildLCL->isImmutable ());
+    assert (buildLCL->info().hash == ledger);
+
+    return RCLCxLedger(buildLCL);
+}
+
+
+std::vector<RCLCxPeerPos>
+RCLConsensus::proposals (LedgerHash const& prevLedger)
+{
+    std::vector <RCLCxPeerPos> ret;
+    {
+        std::lock_guard <std::mutex> _(peerPositionsLock_);
+
+        for (auto const& it : peerPositions_)
+            for (auto const& pos : it.second)
+                if (pos->proposal().prevLedger() == prevLedger)
+                    ret.emplace_back (*pos);
```

Comment:
On further review, the current implementation is also counting this type of copy: https://github.com/ripple/rippled/blob/develop/src/ripple/app/ledger/impl/ConsensusImp.cpp#L158
So I'll leave alone for now.